client/js/contextMenuFactory.js

```diff
@@ -0,0 +1,204 @@
+"use strict";
+const $ = require("jquery");
+const socket = require("./socket");
+const utils = require("./utils");
+const JoinChannel = require("./join-channel");
+const ContextMenu = require("./contextMenu");
+const contextMenuActions = [];
+const contextMenuItems = [];
+
+module.exports = {
+	addContextMenuItem,
+	createContextMenu,
+};
+
+addDefaultItems();
+
+/**
+ *
+ * @param opts
+ * @param {function(Object)} [opts.check] - Function to check whether item should show on the context menu, called with the target jquery element, shows if return is truthy
+ * @param {string|function(Object)} opts.className - class name for the menu item, should be prefixed for non-default menu items (if function, called with jquery element, and uses return value)
+ * @param {string|function(Object)} opts.data - data that will be sent to the callback function (if function, called with jquery element, and uses return value)
+ * @param {string|function(Object)} opts.display - text to display on the menu item (if function, called with jquery element, and uses return value)
+ * @param {function(Object)} opts.callback - Function to call when the context menu item is clicked, called with the data requested in opts.data
+ */
+function addContextMenuItem(opts) {
+	opts.check = opts.check || (() => true);
+	opts.actionId = contextMenuActions.push(opts.callback) - 1;
+	contextMenuItems.push(opts);
+}
+
+function addContextDivider(opts) {
```

Comment:
You don't export `addContextDivider`, is that on purpose?

Also, if I understand correctly, adding a divider can be done from a package with `addContextMenuItem({ /* … */, divider: true /* … */ })`, that's not something we want, right?

This is internal and not API anymore so no big deal, but we could have different classes for these things, and the `items.push` stores those objects directly. `instanceof` to know what type it is, or the constructor of those objects sets a `type` property, etc., but that could help with separation of concerns.

Reply:
> You don't export addContextDivider, is that on purpose?

I didn't do it because it's not used elsewhere, so IDEs complain :-P we can easily add it when we add the package API functionality or call it from elsewhere.

> Also, if I understand correctly, adding a divider can be done from a package

Yeah, that could be done, however it's not like it'll break someone else's code. It's something that someone can do but it would just be them using an undocumented internal API, which we can't always get rid of. 

> This is internal and not API anymore so no big deal, but we could have different classes for these things, and the items.push stores those objects directly. instanceof to know what type it is, or the constructor of those objects sets a type property, etc., but that could help with separation of concerns.

I'm not a real fan of using instanceof because it feels a bit hacky. If we were to do this (we don't have classes at all for the elements) then it would be better to get the classes to generate the html and we just call the function for them to do that. 

I'll have a play around, see what I can do.

Reply:
> we can easily add it when we add the package API functionality or call it from elsewhere.

Yeah that's fine. I was just checking it was on purpose and not forgotten as part of the change.

> I'm not a real fan of using instanceof because it feels a bit hacky.

Over the years my opinion about using `instanceof` has gone back and forth, now I'm like heh, if it does the job 🤷‍♂️ but I get where you're coming from, and I was requesting to use that specifically.

Idea behind it is to separate what is an "Item" and what is a "Divider" (and whatever comes next) so that, among other things, API doesn't leak one into another.
But as I mentioned, it's internal so I don't really care at this point. And I would prefer if we can make sure package authors cannot create a divider like this, but as you said, undocumented API, you're on your own.

Reply:
Yeah, let's leave this out of this PR, but I agree it would be a nice way to go eventually.